Libraries/LibMarkdown/Table.cpp

```diff
@@ -0,0 +1,221 @@
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <LibMarkdown/Table.h>
+
+namespace Markdown {
+
+String Table::render_for_terminal(size_t view_width) const
+{
+    auto segment_width_proportion = view_width == 0 ? 4 : ((float)(view_width - m_columns.size()) / (float)m_total_width);
+    StringBuilder format_builder;
+    StringBuilder builder;
+
+    auto write_aligned = [&](const auto& text, auto width, auto alignment) {
+        size_t original_length = 0;
+        for (auto& span : text.spans())
+            original_length += span.text.length();
+        auto string = text.render_for_terminal();
+        format_builder.clear();
+        if (alignment == Alignment::Center) {
+            auto padding_length = (width - original_length) / 2;
+            builder.appendf("%*s%s%*s", padding_length, "", string.characters(), padding_length, "");
+            if ((width - original_length) % 2)
+                builder.append(' ');
+        } else {
+            format_builder.appendf("%%%s%zus", alignment == Alignment::Left ? "-" : "", width + (string.length() - original_length));
+            builder.appendf(
+                format_builder.to_string().characters(),
+                string.characters());
+        }
+    };
+
+    bool first = true;
+    for (auto& col : m_columns) {
+        if (!first)
+            builder.append('|');
+        first = false;
+        size_t width = col.relative_width * segment_width_proportion;
+        write_aligned(col.header, width, col.alignment);
+    }
+
+    builder.append("\n");
+    for (size_t i = 0; i < view_width; ++i)
+        builder.append('-');
+    builder.append("\n");
+
+    for (size_t i = 0; i < m_row_count; ++i) {
+        bool first = true;
+        for (auto& col : m_columns) {
+            if (col.rows.size() <= i)
+                continue;
+            auto& cell = col.rows[i];
+
+            if (!first)
+                builder.append('|');
+            first = false;
+
+            size_t width = col.relative_width * segment_width_proportion;
+            write_aligned(cell, width, col.alignment);
+        }
+        builder.append("\n");
+    }
+
+    return builder.to_string();
+}
+
+String Table::render_to_html() const
+{
+    StringBuilder builder;
+
+    builder.append("<table>");
+    builder.append("<thead>");
+    builder.append("<tr>");
+    for (auto& column : m_columns) {
+        builder.append("<th>");
+        builder.append(column.header.render_to_html());
+        builder.append("</th>");
+    }
+    builder.append("</tr>");
+    builder.append("</thead>");
+    builder.append("<tbody>");
+    for (size_t i = 0; i < m_row_count; ++i) {
+        builder.append("<tr>");
+        for (auto& column : m_columns) {
+            if (column.rows.size() > i) {
```

Comment:
I'd write this as
```suggestion
            if (i < column.rows.size()) {
```
(pronounced "if `i` is a valid index into `column.rows`)
and perhaps add a comment that columns can have different numbers of rows.

Reply:
Hmmm...actually, this might be a tad bit wrong; it'll shift all empty cells to the end of the column.